src/cargo/core/compiler/mod.rs

```diff
@@ -629,19 +631,42 @@ fn rustdoc<'a, 'cfg>(cx: &mut Context<'a, 'cfg>, unit: &Unit<'a>) -> CargoResult
     }))
 }
 
-fn add_crate_versions_if_requested(
-    bcx: &BuildContext<'_, '_>,
-    unit: &Unit<'_>,
+fn add_crate_versions_if_requested<'a>(
+    bcx: &BuildContext<'a, '_>,
+    unit: &Unit<'a>,
     rustdoc: &mut ProcessBuilder,
 ) {
-    if !bcx.config.cli_unstable().crate_versions {
-        return;
+    if crate_versions_requested(bcx)
+        && !crate_version_flag_in_rustdocflags(bcx, unit)
+        && !crate_version_flag_in_extra_args(bcx, unit)
+    {
+        append_crate_version_flag(unit, rustdoc);
     }
+}
+
+fn crate_versions_requested(bcx: &BuildContext<'_, '_>) -> bool {
+    bcx.config.cli_unstable().crate_versions
+}
+
+fn crate_version_flag_in_rustdocflags(bcx: &BuildContext<'_, '_>, unit: &Unit<'_>) -> bool {
+    bcx.rustdocflags_args(unit)
+        .iter()
+        .any(|flag| flag == RUSTDOC_CRATE_VERSION_FLAG)
```

Comment:
I think these checks will need to be `starts_with` since you can have something like `--crate-version=foo`.